core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogUtils.java

```diff
@@ -85,16 +89,23 @@ public static BrooklynClassLoadingContext getClassLoadingContext(Entity entity)
     }
 
     public static BrooklynClassLoadingContext newClassLoadingContext(@Nullable ManagementContext mgmt, String catalogItemId, Collection<? extends OsgiBundleWithUrl> libraries) {
+        return newClassLoadingContext(mgmt, catalogItemId, libraries, null);
+    }
+    
+    public static BrooklynClassLoadingContext newClassLoadingContext(@Nullable ManagementContext mgmt, String catalogItemId, Collection<? extends OsgiBundleWithUrl> libraries, BrooklynClassLoadingContext loader) {
         BrooklynClassLoadingContextSequential result = new BrooklynClassLoadingContextSequential(mgmt);
 
         if (libraries!=null && !libraries.isEmpty()) {
             result.add(new OsgiBrooklynClassLoadingContext(mgmt, catalogItemId, libraries));
         }
 
-        BrooklynClassLoadingContext loader = BrooklynLoaderTracker.getLoader();
-        if (loader != null) {
+        if (loader !=null) {
```

Comment:
Shouldn't be allowed to stack loaders. A new loader barrier starts at a catalog item and the same loader is used for all types in the item, including specs referenced in URLs.
A loader delegates to a single item's osgi bundles, falling back to the catalog class loader. The root application loader uses only the catalog class loader.
I believe that even adding the thread local loader is wrong (in the sense that shouldn't be needed, but added just in case).